drivers: Add external address translation module and TI RAT support

KarthikL1729 opened the pull request:
This PR adds support for external address translation modules by using the `sys_mm_drv_page_phys_get()` function in the system_mm API. Texas Instruments SoCs (specifically the K3 family) require a Region based Address Translation (RAT) module to be able to access peripherals such as UART, and this PR includes the driver for RAT as well, which utilises the aforementioned API.

Signed off by: <l-lakshmanan@ti.com>

DhruvaG2000 commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,105 @@
+ */
+typedef struct address_trans_region_config_ {
+
+	uint64_t system_addr;	/* translated 48b system addr as seen by the SOC, MUST align to
```
if this is going to be a generic driver, then does it really make sense to say it will be a "48b system addr" always?
I thought this was the case only in TI AM625 M4 core's RAT.

If in future a new device wanted to use RAT does it necessarily have to be a 48bit addr or it can be anything else too?

KarthikL1729 replied:
It can be anything as long as the mapping function accommodates for it, and the mapping function can be custom too. Will remove the comment.

carlocaione reviewed (changes requested):
I don't think this is the way we should go.

Either we fix / adapt the current `system_mm` API and we add a new RAT driver using this API so that we can simply use `sys_mm_drv_map_region()` in `device_mmio.h` or we add an entirely new class for this use-case, outside the `system_mm` API (and directories).

@dcpleung the `system_mm` is a bit of an odd-ball, is this a use case that you envisioned for this API?

KarthikL1729 commented:
> 

RAT is a TI specific module that a few TI SoCs require to be able to access peripherals due to the change in bus size from a 48 bit SoC address to a 32 bit local address (for example in the Cortex M4F in the AM62X SoC).

povergoing commented:
> > 
> 
> RAT is a TI specific module that a few TI SoCs require to be able to access peripherals due to the change in bus size from a 48 bit SoC address to a 32 bit local address (for example in the Cortex M4F in the AM62X SoC).

So, it translates the accesses from the core to the peripherals, right? The initiator is still core? Like, add another region based translation layer under MMU/MPU?

carlocaione commented:
> So, it translates the accesses from the core to the peripherals, right? The initiator is still core? Like, add another region based translation layer under MMU/MPU?

Yes, but this must me somehow integrated into our current MMIO API. Also, we could do that but we do not currently have a proper MMU/MPU API so that should be again some custom / soc-specific code so I was wondering whether we could instead leverage the `system_mm` API for this case to have a proper abstraction layer.

KarthikL1729 commented:
> > > 
> > 
> > 
> > RAT is a TI specific module that a few TI SoCs require to be able to access peripherals due to the change in bus size from a 48 bit SoC address to a 32 bit local address (for example in the Cortex M4F in the AM62X SoC).
> 
> So, it translates the accesses from the core to the peripherals, right? The initiator is still core? Like, add another region based translation layer under MMU/MPU?

The initiator is the core, yes. However, the main requirement in the case of RAT is being able to get a certain local address when given a physical address using the mapping function. 

"It translates a 32-bit input address into a 48-bit output address. Any input transaction that starts inside of a programmed region will have its address translated, if the region is enabled." -> This is from the TRM linked in the commit [here](https://github.com/zephyrproject-rtos/zephyr/pull/60028/commits/2b9577e61e797f1f649caea432c07e779de4d13f).

It works on a transaction basis, and hence, a function like `z_get_local_addr` is what it requires? I was not able to find another function that could serve the same purpose in the current API, and so went ahead with this.

povergoing commented:
> The initiator is the core, yes. However, the main requirement in the case of RAT is being able to get a certain local address when given a physical address using the mapping function.
> 
> "It translates a 32-bit input address into a 48-bit output address. Any input transaction that starts inside of a programmed region will have its address translated, if the region is enabled." -> This is from the TRM linked in the commit [here](https://github.com/zephyrproject-rtos/zephyr/pull/60028/commits/2b9577e61e797f1f649caea432c07e779de4d13f).
> 
> It works on a transaction basis, and hence, a function like `z_get_local_addr` is what it requires? I was not able to find another function that could serve the same purpose in the current API, and so went ahead with this.

Thank you, I got it now.

povergoing commented:
> > So, it translates the accesses from the core to the peripherals, right? The initiator is still core? Like, add another region based translation layer under MMU/MPU?
> 
> Yes, but this must me somehow integrated into our current MMIO API. Also, we could do that but we do not currently have a proper MMU/MPU API so that should be again some custom / soc-specific code so I was wondering whether we could instead leverage the `system_mm` API for this case to have a proper abstraction layer.

My only concern is, if there will be an issue when this RAT and MMU exist at the same time? And will RAT be configured dynamically? I mean configuring it at runtime, mapping the same 32-bit address to 2 devices alternatively?

dcpleung commented:
> @dcpleung the `system_mm` is a bit of an odd-ball, is this a use case that you envisioned for this API?

That API is for interfacing with external MMU/MPU-like hardware that is not architecture specific (e.g. MMU/MPU for ARM and x86).

carlocaione commented:
> > @dcpleung the `system_mm` is a bit of an odd-ball, is this a use case that you envisioned for this API?
> 
> That API is for interfacing with external MMU/MPU-like hardware that is not architecture specific (e.g. MMU/MPU for ARM and x86).

So I guess it is a good match for the RAT then?

dcpleung commented:
Hm... from what I can skim through the doc... for example, K3 is based on Cortex-R5, which has MPU. How would the RAT module and MPU work together? Unless you add a `depends on !MPU` on the Kconfig, there is always a chance someone will enable them together.

Correct me if I am wrong (as I have only skimmed through the doc), RAT module seems like a limited scope MMU where you can map some (limited) memory region to another memory region (so... "translating" the address from one to another). However, there is no access control, and this is pure address translation. It would help to know what kind of use cases are out there, which would help us to review this.

dcpleung commented:
> > > @dcpleung the `system_mm` is a bit of an odd-ball, is this a use case that you envisioned for this API?
> > That API is for interfacing with external MMU/MPU-like hardware that is not architecture specific (e.g. MMU/MPU for ARM and x86).
> So I guess it is a good match for the RAT then?

Possibly... I would like to see use cases to help me understand the bigger picture.

vaishnavachath commented:
@dcpleung @povergoing @carlocaione , Thank you for the review, feedback, adding some notes to address your questions:
 
@povergoing 
> 
> My only concern is, if there will be an issue when this RAT and MMU exist at the same time? And will RAT be configured dynamically? I mean configuring it at runtime, mapping the same 32-bit address to 2 devices alternatively?

RAT will be primarily be used by the M4 cores in the K3 SoCs where there is no MMU available, but from a hardware perspective the R5 in these devices which has MPU can also make use of RAT, but there is no use case where both MPU/MMU and RAT is configured at same time and we would like to support use cases where RAT is enabled independently like for M4. From hardware perspective dynamic reconfiguration of RAT is possible, but again this is also not a valid use case we have seen in the past.


@dcpleung 

> Hm... from what I can skim through the doc... for example, K3 is based on Cortex-R5, which has MPU. How would the RAT module and MPU work together? Unless you add a `depends on !MPU` on the Kconfig, there is always a chance someone will enable them together.
> 
> Correct me if I am wrong (as I have only skimmed through the doc), RAT module seems like a limited scope MMU where you can map some (limited) memory region to another memory region (so... "translating" the address from one to another). However, there is no access control, and this is pure address translation. It would help to know what kind of use cases are out there, which would help us to review this.

Just clarifying a few pieces, your description is correct, RAT can be thought of as a limited scope MMU which performs only address translation.
K3 (Keystone 3) SoCs is a family of heterogeneous multi-core application processors from TI which usually contains multicore ARM64 application processors (A53, A72), co-processors(R5, M4), DSPs .etc. The device under discussion here AM62 has Quad A53 cores and a single M4 core for which @KarthikL1729 is trying to add support for here: https://github.com/zephyrproject-rtos/zephyr/pull/58722 and the RAT support is a dependency as the M4 does not have a MMU/MPU and the core needs to access memory region outside the 32-bit memory space.

This FAQ also describes the primary use cases with RAT : https://e2e.ti.com/support/processors-group/processors/f/processors-forum/1192281/faq-am62x-am64x-updating-the-region-based-address-translation-rat-settings

Looking at the system_mm description you and @carlocaione provided and https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/mm/system_mm.h#L12 , it seems like that is where the RAT implementation belongs, please let us know if the understanding is correct.

carlocaione commented:
> Looking at the system_mm description you and @carlocaione provided and https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/mm/system_mm.h#L12 , it seems like that is where the RAT implementation belongs, please let us know if the understanding is correct.

In my opinion (not sure what @dcpleung thinks) but you can use the `system_mm` API to abstract RAT away.

dcpleung commented:
It would be great if you can make it work with the existing `system_mm` API.

carlocaione commented on include/zephyr/sys/mem_manage.h:
```diff
@@ -261,6 +261,26 @@ void z_phys_map(uint8_t **virt_ptr, uintptr_t phys, size_t size,
+
+__weak void *z_get_local_addr(uint64_t phys_addr);
+#endif /* CONFIG_EXTERNAL_ADDRESS_TRANSLATION */
+
```
Why this is being introduced here and removed in the 3rd commit?

carlocaione commented on drivers/mm/Kconfig:
```diff
@@ -45,3 +45,10 @@ config MM_DRV_INTEL_ADSP_TLB
 
 endif # MM_DRV
+
+menuconfig EXTERNAL_ADDRESS_TRANSLATION
```
why this is menuconfig? and why this is here at all?

carlocaione commented on drivers/mm/rat/CMakeLists.txt:
this is an `mm` driver, no need to have one new `rat` sub-directory for it.

carlocaione commented on drivers/mm/rat/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+zephyr_library()
+
+zephyr_library_sources(ti_rat.c)
```
not needed, just use `drivers/mm/CMakeLists.txt`

carlocaione commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,100 @@
+/**
+ * \brief Enum's to represent different possible region size for the address translate module
+ */
+typedef enum address_trans_region_size_e {
```
please do not use `typedefs`

carlocaione commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,100 @@
+#define RAT_BASE(base_addr, i)     (volatile uint32_t *)(base_addr + 0x24 + 0x10 * (i))
+#define RAT_TRANS_L(base_addr, i)  (volatile uint32_t *)(base_addr + 0x28 + 0x10 * (i))
+#define RAT_TRANS_H(base_addr, i)  (volatile uint32_t *)(base_addr + 0x2C + 0x10 * (i))
+#ifndef NULL
+#define NULL (0)
+#endif
```
why?

carlocaione commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,100 @@
+void RAT_deinit(void);
+void *rat_get_local_addr(uint64_t system_addr);
+/* Generic abstraction function to override weakly defined function */
+void *z_get_local_addr(uint64_t system_addr);
```
Why are you exporting this interface? The only interface exported should be the `system_mm` API.

carlocaione commented on drivers/mm/rat/ti_rat.c:
```diff
@@ -0,0 +1,111 @@
+address_trans_params translate_config;
+
+void address_trans_set_region(uint32_t rat_base_addr, uint16_t region_num, uint64_t system_addr,
+			      uint32_t local_addr, uint32_t size, uint32_t enable)
```
all the functions not related to the API should be `static`.

carlocaione commented on drivers/mm/rat/ti_rat.c:
```diff
@@ -0,0 +1,111 @@
+	*RAT_BASE(rat_base_addr, region_num) = local_addr;
+	*RAT_TRANS_L(rat_base_addr, region_num) = system_addrL;
+	*RAT_TRANS_H(rat_base_addr, region_num) = system_addrH;
+	*RAT_CTRL(rat_base_addr, region_num) = ((enable & 0x1) << 31u) | (size & 0x3F);
```
Use the proper `sys_write*` helpers for this.

carlocaione commented on drivers/mm/CMakeLists.txt:
```diff
@@ -16,4 +16,4 @@ zephyr_sources_ifdef(
 	)
 
-add_subdirectory_ifdef(CONFIG_TI_RAT rat)
+zephyr_sources_ifdef(CONFIG_TI_RAT mm_drv_ti_rat.c)
```
why are you modifying files that you created literally in the previous commit.

carlocaione reviewed (changes requested):
Please, properly squash the commits.

KarthikL1729 commented:
@carlocaione Apologies, I pushed the new files from my local system without squashing, hence the multiple changes in files created in this commit. Will rebase and push.

KarthikL1729 committed:
drivers: mm: Add support for TI RAT module using system_mm API

Added Region based Address Translation (RAT) module driver. Required by
a few Texas Instruments SoCs to fucntion. Uses
sys_mm_drv_page_phys_get() API with device_map() for address translation.

Signed-off-by: L Lakshmanan <l-lakshmanan@ti.com>

KarthikL1729 commented:
I have made the changes to use the system_mm API, and this implementation should be functional. It is all in one commit as there is no extension to an API now. @carlocaione I will rebase and push a fix for the typedefs after implementing a workaround.

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,117 @@
+	system_addrH = (uint32_t)((system_addr >> 32) & 0xFFFF);
+	local_addr = local_addr & ~((uint32_t)(((uint64_t)1 << size) - 1));
+
+	*RAT_CTRL(rat_base_addr, region_num) = 0;
+	*RAT_BASE(rat_base_addr, region_num) = local_addr;
+	*RAT_TRANS_L(rat_base_addr, region_num) = system_addrL;
+	*RAT_TRANS_H(rat_base_addr, region_num) = system_addrH;
+	*RAT_CTRL(rat_base_addr, region_num) = ((enable & 0x1) << 31u) | (size & 0x3F);
```
Please use the `sys_write*` helpers to access registers.

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,117 @@
+			 "RAT region config cannot be NULL");
+
+		/* enable regions setup by user */
+		address_trans_set_region(translate_config.rat_base_addr, i,
+					 translate_config.region_config[i].system_addr,
+					 translate_config.region_config[i].local_addr,
+					 translate_config.region_config[i].size, 1);
```
can you just pass `translate_config` or `translate_config.region_config`? no need to overload the stack.

KarthikL1729 commented on drivers/mm/CMakeLists.txt:
```diff
@@ -15,3 +15,5 @@ zephyr_sources_ifdef(
 	mm_drv_intel_adsp_mtl_tlb.c
 	)
+
+zephyr_sources_ifdef(CONFIG_TI_RAT mm_drv_ti_rat.c)
```
RAT is still TI specific, so I thought this would be better? Or would CONFIG_MM_TI_RAT be preferred?

carlocaione replied:
`CONFIG_MM_TI_RAT` is better.

KarthikL1729 commented on drivers/mm/Kconfig:
```diff
@@ -44,4 +44,22 @@ config MM_DRV_INTEL_ADSP_TLB
+	help
+	  This config is intended to support an external address
+	  translation module if required for an SoC. Extends the
+	  MMIO API with the z_get_local_addr() function.
```
Sorry, overlooked this. Will change.

povergoing commented on include/zephyr/sys/device_mmio.h:
```diff
@@ -101,8 +105,11 @@ static inline void device_map(mm_reg_t *virt_addr, uintptr_t phys_addr,
 	ARG_UNUSED(flags);
-
+#ifdef CONFIG_EXTERNAL_ADDRESS_TRANSLATION
+	sys_mm_drv_page_phys_get((void *) phys_addr, virt_addr);
```
I have no objection to this since this is the only use case for now?

Just raise my concern for the future as this is a system API:
1. MMU/MPU and RAT (or alternative technologies) exist at the same time which is hard to handle.
2. Suppose we remove the exclusive macro (like `z_phys_map` firstly, and then `sys_mm_drv_page_phys_get`, it's still annoying that if we want to remap RAT only, however, we inevitably call MMU remap again.
3.  How do we deal with similar techniques like Arm ATU?

Fortunately, we could think about it later. MMU and RAT existing simultaneously seem like a 2 stage translation which is definitely a tricky thing.

carlocaione commented on drivers/mm/Kconfig:
```diff
@@ -44,4 +44,22 @@ config MM_DRV_INTEL_ADSP_TLB
 	  Driver for the translation lookup buffer on
 	  Intel Audio DSP hardware.
 
+config EXTERNAL_ADDRESS_TRANSLATION
+	bool "Support for external address translation modules if necessary"
+	help
+	  This config is intended to support an external address
+	  translation module if required for an SoC. Uses the
+	  sys_mm_drv_page_phys_get() function from the system_mm API.
```
This should depends on `!MMU`

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,118 @@
+
+struct address_trans_params translate_config;
+
+static void address_trans_set_region(struct address_trans_params addr_translate_config,
```
use a pointer here, just be nice on the stack.

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,118 @@
+	sys_write32(local_addr, RAT_BASE(rat_base_addr, region_num));
+	sys_write32(system_addrL, RAT_TRANS_L(rat_base_addr, region_num));
+	sys_write32(system_addrH, RAT_TRANS_H(rat_base_addr, region_num));
+	sys_write32(((enable & 0x1) << 31u) | (size & 0x3F), RAT_CTRL(rat_base_addr, region_num));
```
use `defines` for the hardcoded numerical values.

carlocaione commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,92 @@
+ * \brief Parameters for \ref address_trans_init
+ */
+struct address_trans_params {
+
```
nit: empty line

carlocaione commented on include/zephyr/drivers/mm/rat.h:
```diff
@@ -0,0 +1,92 @@
+ * \brief Region config structure
+ */
+struct address_trans_region_config {
+
```
nit: empty line

carlocaione commented on include/zephyr/sys/device_mmio.h:
```diff
@@ -33,10 +33,14 @@
 #define DEVICE_MMIO_IS_IN_RAM
 #endif
 
+#if defined(CONFIG_MM_DRV)
```
why not `CONFIG_EXTERNAL_ADDRESS_TRANSLATION`?

carlocaione commented on drivers/mm/Kconfig:
```diff
@@ -44,4 +44,23 @@ config MM_DRV_INTEL_ADSP_TLB
 	  Intel Audio DSP hardware.
 
+config EXTERNAL_ADDRESS_TRANSLATION
+	bool "Support for external address translation modules if necessary"
```
```suggestion
	bool "Support for external address translation modules"
```

carlocaione commented on drivers/mm/Kconfig:
```diff
@@ -44,4 +44,23 @@ config MM_DRV_INTEL_ADSP_TLB
+	bool "Texas Instruments RAT module"
+	depends on EXTERNAL_ADDRESS_TRANSLATION
+	help
+	  Enables Region based address translation
+	  support functions specific to TI SoCs.
```
nit: re-indent this.

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,144 @@
+	if (size > address_trans_region_size_4G) {
+		size = address_trans_region_size_4G;
+	}
+	system_addrL = (uint32_t)(system_addr & ~((uint32_t)(((uint64_t)1 << size) - 1)));
```
Can you use `BIT_MASK()` here?

carlocaione commented on drivers/mm/mm_drv_ti_rat.c:
```diff
@@ -0,0 +1,144 @@
+		uint32_t size_mask;
+
+		size_mask =
+			((uint32_t)(((uint64_t)1 << translate_config.region_config[regionId].size) -
```
same here `BIT_MASK()`

carlocaione reviewed (dismissed):
Just a couple of minor comments.
